src/Compilers/CSharp/Portable/Compilation/DeconstructionInfo.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    // Deconstructions are represented internally as a tree of conversions, but
+    // since they are not conversions from the language perspective we use a wrapper to
+    // abstract the public API from the implementation.
+    public struct DeconstructionInfo : IEquatable<DeconstructionInfo>
+    {
+        private readonly Conversion _conversion;
+
+        public IMethodSymbol Method
+        {
+            get
+            {
+                return _conversion.Kind == ConversionKind.Deconstruction
+                    ? _conversion.Method
+                    : null;
+            }
+        }
+
+        public Conversion? Conversion
+        {
+            get
+            {
+                return _conversion.Kind == ConversionKind.Deconstruction
+                    ? null
+                    : (Conversion?)_conversion;
+            }
+        }
+
+        public ImmutableArray<DeconstructionInfo> Nested
+        {
+            get
+            {
+                var underlyingConversions = _conversion.UnderlyingConversions;
+
+                Debug.Assert(!underlyingConversions.IsDefault || _conversion.Kind != ConversionKind.Deconstruction);
+
+                return underlyingConversions.IsDefault
+                    ? ImmutableArray<DeconstructionInfo>.Empty
+                    : underlyingConversions.SelectAsArray(c => new DeconstructionInfo(c));
+            }
+        }
+
+        internal DeconstructionInfo(Conversion conversion)
+        {
+            _conversion = conversion;
```

Comment:
Assert here that conversion.Kind is Deconstruction since everything else is presuming that?

Reply:
Only non-terminal nodes have kind=Deconstruction. The leaf nodes are regular conversions (Identity, Numeric, etc).